src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,152 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDCheckInSender.h"
+
+#include <app/InteractionModelEngine.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <protocols/secure_channel/CheckinMessage.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+namespace app {
+
+using namespace Protocols::SecureChannel;
+
+ICDCheckInSender::ICDCheckInSender(PersistentStorageDelegate * storage, FabricTable * fabricTable,
+                                   Crypto::SymmetricKeystore * symmetricKeyStore)
+{
+    VerifyOrDie(storage != nullptr);
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+
+    mAddressLookupHandle.SetListener(this);
+}
+
+ICDCheckInSender::ICDCheckInSender()
+{
+    mAddressLookupHandle.SetListener(this);
+}
+
+void ICDCheckInSender::Init(PersistentStorageDelegate * storage, FabricTable * fabricTable,
+                            Crypto::SymmetricKeystore * symmetricKeyStore)
+{
+    VerifyOrDie(storage != nullptr);
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+}
+
+void ICDCheckInSender::OnNodeAddressResolved(const PeerId & peerId, const AddressResolve::ResolveResult & result)
+{
+    mResolveInProgress = false;
+
+    ICDMonitoringEntry entry(mSymmetricKeystore);
+
+    const FabricInfo * f = mFabricTable->FindFabricWithCompressedId(peerId.GetCompressedFabricId());
+    ICDMonitoringTable table(*mStorage, f->GetFabricIndex(), CHIP_CONFIG_ICD_CLIENTS_SUPPORTED_PER_FABRIC, mSymmetricKeystore);
+    table.Find(peerId.GetNodeId(), entry);
+
+    if (CHIP_NO_ERROR != SendCheckInMsg(entry, result.address))
+    {
+        ChipLogError(AppServer, "Failed to send the ICD Check-In message");
+    }
+}
+
+void ICDCheckInSender::OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason)
+{
+    mResolveInProgress = false;
+    ChipLogProgress(AppServer, "Node Address resolution failed for ICD Check-In with Node ID " ChipLogFormatX64,
+                    ChipLogValueX64(peerId.GetNodeId()));
+}
+
+CHIP_ERROR ICDCheckInSender::SendCheckInMsg(ICDMonitoringEntry & entry, const Transport::PeerAddress & addr)
+{
+    uint8_t b[CheckinMessage::sMinPayloadSize] = { 0 };
+    MutableByteSpan output{ b };
+    CHIP_ERROR err;
```

Comment:
```suggestion
    CHIP_ERROR err = CHIP_NO_ERROR;
```
For good practice, we should not have an undefined variable.